Add export campaign feature; Closes #1849

thefoxwithin opened the pull request:
Please ensure you are familiar with our Pull Request Description expectations described here: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/
### What?
Add the ability to export a full campaign (category) and all its quests from a tenant schema to the shared library schema, including confirmation views and permission checks.

### Why?
This feature enables users to share campaigns and their quests centrally in the shared library, improving content reuse and management across tenants. 

https://github.com/bytedeck/bytedeck/issues/1849

### How?
- Implement `ExportCampaignView` with GET confirmation and POST export logic.
- Add permission checks to restrict export to deck owners or staff if `allow_staff_export=True`.
- Use import/export logic with `QuestResource` to transfer campaigns and quests.
- Notify staff users after a successful export.
- Add tests covering permission enforcement, export success, and error cases.

### Testing?
- Unit tests verify export permission scenarios for deck owners and staff.
- Tests ensure forbidden access when a campaign already exists in the library.
- Tests confirm successful export redirects and notifications are sent.
- Manual testing of confirmation pages and export flows in UI.

### Screenshots (if front end is affected)

#### With no Conflicts
<img width="825" height="747" alt="image" src="https://github.com/user-attachments/assets/b345b08b-0aff-4f88-9f2e-d21cc4686b08" />

#### One or more Quests exist in the Library already
<img width="841" height="805" alt="image" src="https://github.com/user-attachments/assets/72fe541d-6879-4fcd-8478-735340a4f4b9" />

#### Campaign already exists in the Library already
<img width="1128" height="292" alt="image" src="https://github.com/user-attachments/assets/7bbe1973-3339-4e63-9e55-c106547ab0c9" />

#### Success notification displayed on Library
<img width="740" height="162" alt="image" src="https://github.com/user-attachments/assets/820ab3e0-9667-4e6d-9a4a-86ff9cb19a82" />

> The success notification uses the same format as the quests, Library `deck_ai`, user who exported, the deck it was exported from, and finally the name of the campaign. Clicking the notification will send you to the Campaign detail page

### Anything Else?
- Future improvements include handling duplicate quests in library by creating copies.


### Review request
@tylerecouture


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Export entire campaigns (category + published quests) to the shared library with a confirmation page, license acceptance, conflict warnings, routing, and post-export notifications.

- Improvements
  - Centralized export permission checks; clearer conflict warnings and library indicators in lists; exported items are created unpublished in the library; isolated cross-tenant operations; refined redirects and success messaging.

- Tests
  - Expanded coverage for permissions, conflict handling, successful quest/campaign exports, notifications, and cross-tenant scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tylerecouture commented on src/library/exporter.py:
```diff
@@ -52,3 +52,59 @@ def export_quest_to_library(*, source_schema, quest_import_id):
+
+        # Unpublish exported quests
+        exported_quests = Quest.objects.filter(import_id__in=[q.import_id for q in quests])
+        for eq in exported_quests:
```
Code comment with context for why this loop is needed.   Will staff get a notification for EVERY quest in the campaign that is sent to the library?  Maybe we don't need signals to trigger on this on the library?  Need some more comments why we are doing it this way and what the consequences are.

tylerecouture replied:
I'm curious how much lag there is when doing this for real, since it's not asynchronous, how slow is this process?  If I export a campaign of 20 quests, does it take 2 seconds or 10 or 30?

thefoxwithin replied:
With 31 quests it took a little over 6 seconds to complete. Approximately 6.3 to be more precise, which works out to about 0.20 seconds per quest. That seems like a reasonable ballpark, but let me know if that feels acceptable.

I also tested the scenario where every quest already existed on the library. In that case, the export took about 6.77 seconds, a small increase that works out to 0.21 seconds per quest. This suggests the existence check is only a minor additional cost.

To see how this scales, I added 19 more quests (total of 50), with 31 of them already present on the library. The export finished in 10.44 seconds, which is consistent with the ~0.21 seconds per quest observed earlier.

As an experiment, I also tried switching the “unpublish quests” step from looping through each quest and calling .save() to a single bulk update that skips model validation and signals. Surprisingly, this was slower, about 11.40 seconds for the same 50-quest export. On investigation, this makes sense: the loop was already very cheap (just one boolean flip and update per row), and the real bottleneck in export is the serialization/import logic and database writes, not this small unpublish step. A bulk update might even incur extra overhead if the published field is indexed, since PostgreSQL has to touch all affected rows at once. So the performance difference here is negligible, and the choice between bulk update vs loop is more about clarity and signal handling than speed.

> Will staff get a notification for EVERY quest in the campaign that is sent to the library?

No, only a single notification for the campaign itself is sent. Clicking it takes staff to the campaign page where they can see all the included quests.

Looking at it a bit closer the exporter and importer are doing it differently, I'll change the exporter to act more like the importer and this should remove the need to loop over all the quests. (The importer doesn't even do that which is what tipped me off that they're behaving differently)

thefoxwithin replied:
After updating it to be a lot more like the importer it appears there is no change in speed.

still about 10.5 seconds for exporting a campaign with 50 quests.

tylerecouture replied:
Anything that takes more than a few seconds should be done asychronously as a celery task.  I don't think you've done any of these yet, so might as well learn it!

You'll need to update the message so that it lets the user know the quests are being pushed to the library in the background.

Create a celery tasks and run it.  The user will have their next page load right away and the actually processing will be done concurrently.

tylerecouture replied:
Alternately, you an just leave the message as is, is there any difference to the user whether the quests appear instantly in the library or take 10 seconds?

tylerecouture replied:
An example if you want one, is the "regenerate all maps" view `regenerate_all` in djcytoscape/views.py.

It checks if there are more than 5 maps to regenerate, it sends it as a celery task/parallel instead of doing it in series.

tylerecouture replied:
One more thing, is that local development environment might be a lot slower than production, so maybe test it in production first.  If uploading 30 quests takes 4 seconds or less in production, we can leave it as is.

tylerecouture reviewed (changes requested):
A couple comments about efficiency.  Can you test this and how long it takes in testing to export a big campaign of like 30 quests?  try auto generating some big campaigns using the admin tool

thefoxwithin committed:
Add export campaign feature with confirmation page and conflict detection

add in todos

Add comprehensive permission and export tests for campaign library views

added full_clean to tests updated exporter more defensive template check

reduced duplication by making ExportPermissionMixin for the export views

Updated tests to follow naming conventions

Refactor export_campaign_to_library to mirror importer behavior

include archived quests in the check on the library for duplicates

perform export outside of library context

Narrow update for efficiency
